go/plugins/ollama/ollama.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ollama
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/firebase/genkit/go/ai"
+)
+
+const provider = "ollama"
+
+func defineModel(name string, serverAddress string) {
+	meta := &ai.ModelMetadata{
+		Label: "Ollama - " + name,
+		Supports: ai.ModelCapabilities{
+			Multiturn: true,
+		},
+	}
+	g := generator{model: name, serverAddress: serverAddress}
+	ai.DefineModel(provider, name, meta, g.generate)
+}
+
+// Model returns the [ai.ModelAction] with the given name.
+// It returns nil if the model was not configured.
+func Model(name string) *ai.ModelAction {
+	return ai.LookupModel(provider, name)
+}
+
+// Config provides configuration options for the Init function.
+type Config struct {
+	// API key. Required.
+	ServerAddress string
+	// Generative models to provide.
+	Models []string
+}
+
+// Init registers all the actions in this package with ai.
+func Init(ctx context.Context, cfg Config) error {
+	for _, model := range cfg.Models {
+		defineModel(model, cfg.ServerAddress)
+	}
+	return nil
+}
+
+type generator struct {
+	model         string
+	serverAddress string
+}
+
+/*
+TODO: Support optional, advanced parameters:
+format: the format to return a response in. Currently the only accepted value is json
+options: additional model parameters listed in the documentation for the Modelfile such as temperature
+system: system message to (overrides what is defined in the Modelfile)
+template: the prompt template to use (overrides what is defined in the Modelfile)
+context: the context parameter returned from a previous request to /generate, this can be used to keep a short conversational memory
+stream: if false the response will be returned as a single response object, rather than a stream of objects
+raw: if true no formatting will be applied to the prompt. You may choose to use the raw parameter if you are specifying a full templated prompt in your request to the API
+keep_alive: controls how long the model will stay loaded into memory following the request (default: 5m)
+*/
+type ollamaRequest struct {
+	Messages []map[string]string `json:"messages"`
```

Comment:
I'm not a fan of this being just a map[string]string. We should probably define internal types here with the correct fields/subfields and have json labels. Just taking a stab at it:

type ollamaMessage struct {
    role string // `json:"role"`
    content string // `json:"content"`
}

Then ollamaRequest.Messages would have type []*ollamaMessage